scripts/generate-sitemap.js

```diff
@@ -13,63 +13,147 @@ async function readSchemaJson(filePath) {
   }
 }
 
-async function processDirectory(dirPath, relativePath = "") {
-  const entries = await fs.readdir(dirPath, { withFileTypes: true });
-  let paths = [];
-
-  for (const entry of entries) {
-    const entryPath = path.join(dirPath, entry.name);
-
-    if (entry.name === "index.json") {
-      continue;
+async function processFile(fullPath, relativePath, name) {
+  const schemaPath = _path.join(fullPath, name);
+  const schemaData = await readSchemaJson(schemaPath);
+  if (!schemaData) {
+    return null;
+  }
+  const pageName = name.split(".")[0];
+  const permalink = _path.join(relativePath, pageName);
+  const siteMapEntry = {
+    permalink,
+    title:
+      schemaData.page.title ||
+      pageName.charAt(0).toUpperCase() + pageName.slice(1),
+  };
+  // check if file is actually an index page for a directory
+  try {
+    const directoryPath = _path.join(fullPath, name.split(".")[0]);
+    const stats = await fs.stat(directoryPath);
+    if (stats.isDirectory()) {
+      return {
+        ...siteMapEntry,
+        children: await processDirectory(directoryPath, permalink),
+      };
     }
+    return siteMapEntry;
+  } catch (error) {
+    return siteMapEntry;
+  }
+}
 
-    if (entry.isDirectory()) {
-      const subRelativePath = path.join(relativePath, entry.name);
+// Returns new index files that were generated as they might have to be added to _pages.json
+async function generateIndexFilesForDirectoriesIfMissing(dirEntries) {
+  const newIndexFiles = [];
+  for (const dirEntry of dirEntries) {
+    const indexFilePath = _path.join(dirEntry.path, dirEntry.name + ".json");
+    try {
+      await fs.access(indexFilePath);
+    } catch (error) {
+      const dirName = dirEntry.name.split(".")[0].replace(/-/g, " ");
+      await fs.writeFile(
+        indexFilePath,
+        JSON.stringify(
+          {
+            version: "0.1.0",
+            layout: "content",
+            page: {
+              title: dirName.charAt(0).toUpperCase() + dirName.slice(1),
+              contentPageHeader: {},
+            },
+            content: [],
```

Comment:
question: for the content should we generate a list of pages in the folder?

Reply:
oh I left it out as it will be shown in the side rail
can be done, think will need to recurse in order to get its children (as they might be missing the index file also lol)

was actually planning to revisit this when we have a more concrete index page layout (smth simple like the gov.uk ones), since it's kind of an edge case and PR getting a bit cancer

Reply:
nvm, I think should be less cancer if I move the processing into this function and recurse inside also, instead of currently where I only populate the index file then process it outside in the caller

Reply:
[6a285bd](https://github.com/isomerpages/isomer-next-build/pull/1/commits/6a285bd0062151fae66e20fa61fe236b828f65a1)